sklearn/feature_extraction/text.py

```diff
@@ -573,7 +574,7 @@ def _document_frequency(X):
     if sp.isspmatrix_csr(X):
         return np.bincount(X.indices, minlength=X.shape[1])
     else:
-        return np.diff(sp.csc_matrix(X, copy=False).indptr)
+        return np.diff(X.indptr)
```

Comment:
Checked that here we only end-up here with CSC matrices, so the `csc_matrix` is indeed redundant.

Still it's not obvious when reading the code. I would either add a comment, or keep it as it was (I guess it will simply validate sparse CSC array type with `copy=False`).

**Edit:**  this line is not covered, so if we do change it, we would need to add a test.

Reply:
After more thoughts, it might make sense to keep it as it was.

Otherwise, if someone were to provide `scipy.sparse.bsr_matrix` to it (e.g. if the code base changes in the future or someone is using this private method) it will silently produce incorrect results, I think.

There is no real advantage in removing it, short of reduced verbosity, right?

Reply:
That might be slower:

```python
In [5]: X = sparse.random(100000, 1000).tocsc()

In [7]: %timeit np.diff(X.indptr)
3.65 µs ± 42.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [8]: %timeit np.diff(sparse.csc_matrix(X, copy=False).indptr)
33.6 µs ± 3.27 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

```

Reply:
bsr_matrix will be converted by the `check_array`

Reply:
> bsr_matrix will be converted by the check_array

I know, it was more a general comment of not changing something if there is no need to. 

Fair enough, the speedup is significant.